admin_guide/encrypting_data.adoc

```diff
@@ -0,0 +1,291 @@
+| Must be rotated every 200k writes
+| Fastest
+| 16, 24, or 32-byte
+| *Is not recommended* for use except when an automated key rotation scheme is implemented.
+
+|===
+
+Each provider supports multiple keys. The keys are tried in order for decryption, and if the
+provider is the first provider, the first key is used for encryption.
+
+[NOTE]
+====
+Kubernetes has no proper nonce generator and uses a random IV as nonce for AES-GCM. Since AES-GCM
+requires a proper nonce to be secure, AES-GCM is not recommended. The 200k write limit just limits
+the possibility of a fatal nonce misuse to a reasonable low margin.
+====
+
+[[encrypting-data-process]]
+== Encrypting Data
+
+Create a new encryption config file
+
+[source,yaml]
+----
+kind: EncryptionConfig
+apiVersion: v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - aescbc:
+        keys:
+        - name: key1
+          secret: <BASE 64 ENCODED SECRET>
+    - identity: {}
+----
+
+To create a new secret perform the following steps:
+
+. Generate a 32 byte random key and base64 encode it. For example, on Linux and
+macOS the following command can be used:
++
+----
+head -c 32 /dev/urandom | base64
+----
++
+[IMPORTANT]
+====
+The encryption key must be generated with an appropriate cryptographically secure random number
+generator like *_/dev/urandom_*. For example, `math/random` from Golang or `random.random()` from
+Python are not suitable.
+====
+
+. Place that value in the *secret* field
+
+. Restart API server
+
+[IMPORTANT]
+====
+Config file contains keys that can decrypt content in etcd, so you must properly restrict
+permissions on masters so only the user who runs the master API server can read it.
+====
+
+
+[[encrypting-data-verification]]
+== Verifying that Data is Encrypted
+
+Data is encrypted when written to etcd. After restarting API server any newly created or
+updated secrets should be encrypted when stored. To check, you can use the `etcdctl` command line
+program to retrieve the contents of your secret.
+
+. Create a new secret called `secret1` in the `default` namespace:
++
+----
+oc create secret generic secret1 -n default --from-literal=mykey=mydata
+----
+
+. Using the `etcdctl` commandline, read that secret out of etcd:
++
+----
+ETCDCTL_API=3 etcdctl get /kubernetes.io/secrets/default/secret1 -w fields [...] | grep Value
+----
++
+where *[...]* must be the additional arguments for connecting to the etcd server.
++
+The final command may look like this:
++
+----
+ETCDCTL_API=3 etcdctl get /kubernetes.io/secrets/default/secret1 -w fields \
+--cacert=/var/lib/origin/openshift.local.config/master/ca.crt \
+--key=/var/lib/origin/openshift.local.config/master/master.etcd-client.key \
+--cert=/var/lib/origin/openshift.local.config/master/master.etcd-client.crt \
+--endpoints 'https://127.0.0.1:4001' | grep Value
+----
+
+. Verify the stored secret is prefixed with *k8s:enc:aescbc:v1:* which indicates the *aescbc*
+provider has encrypted the resulting data.
```

Comment:
It supposed that the previous command returned a secret as raw data and here we're verifying that this raw representation "is prefixed with *k8s:enc:aescbc:v1:*".